Enable test module teardown by default

crisbeto opened the pull request:
These are the changes necessary to enable the `destroyAfterEach` teardown behavior by default. They're split into the following commits:

**feat(core): enable test module teardown by default**
Sets the `destroyAfterEach` teardown behavior to be enabled by default.

**feat(core): add migration to opt out existing apps from new test module teardown behavior**
Since the `destroyAfterEach` teardown behavior is enabled by default now, existing tests that depended on the old behavior can start to fail. These changes add an automated migration that explicitly adds `destroyAfterEach: false` to existing tests.

The migration works by looking for `initTestEnvironment` calls across the entire app and adding the flag to them. If no calls were found, the migration will add the flag to all `configureTestingModule` and `withModule` calls instead.

crisbeto committed:
feat(core): enable test module teardown by default

Sets the `destroyAfterEach` teardown behavior to be enabled by default.

crisbeto commented on packages/router/test/integration.spec.ts:
```diff
@@ -5854,42 +5853,6 @@ describe('Integration', () => {
        }));
   });
-
-  describe('RouterInitializer', () => {
```
I deleted this test, because I introduced it in #42560 specifically to test an error when test module teardown is enabled. Now that it's enabled by default, we don't need it and it'll throw an error because it'll try to destroy an already-destroyed module.

Alternatively we could keep it by adding `destroyAfterEach: false`, but we'll have to delete it eventually when the `teardown` flag is removed.

crisbeto commented on packages/service-worker/test/comm_spec.ts:
```diff
@@ -469,7 +469,7 @@ import {MockPushManager, MockPushSubscription, MockServiceWorkerContainer, MockS
           });
         });
-        return update.activateUpdate().then(() => done()).catch(err => done.fail(err));
+        update.activateUpdate().then(() => done()).catch(err => done.fail(err));
```
These changes aren't necessary, but I noticed that Jasmine logs a warning for them, because they were using the `done` parameter and returning a promise which isn't supported.

crisbeto commented on packages/core/testing/src/test_bed.ts:
```diff
@@ -451,8 +451,11 @@ export class TestBedViewEngine implements TestBed {
-    (this._moduleRef.injector.get(ApplicationInitStatus) as any).runInitializers();
-    this._instantiated = true;
+    try {
+      (this._moduleRef.injector.get(ApplicationInitStatus) as any).runInitializers();
```
I found this issue while debugging a test failure in `service-worker`. We have a test that triggers an error inside the app initializer on purpose which in turn was preventing the VE version of `TestBed` from being cleaned up correctly, causing other tests to fail.

atscott commented on packages/core/schematics/migrations/testbed-teardown/index.ts:
```diff
@@ -0,0 +1,75 @@
+/** Migration that adds the `teardown` flag to `TestBed` calls. */
+export default function(): Rule {
+  return (tree: Tree) => {
+    const {buildPaths, testPaths} = getProjectTsConfigPaths(tree);
```
FYI, this method is now async - https://github.com/angular/angular/commit/2efc18e6757157589004e23d8d22b7967de4387d

atscott reviewed (approved):
reviewed-for: fw-router

jessicajaniuk reviewed (approved):
LGTM :cookie:

crisbeto committed:
feat(core): add migration to opt out existing apps from new test module teardown behavior

Since the `destroyAfterEach` teardown behavior is enabled by default now, existing tests that depended on the old behavior can start to fail. These changes add an automated migration that explicitly adds `destroyAfterEach: false` to existing tests.

The migration works by looking for `initTestEnvironment` calls across the entire app and adding the flag to them. If no calls were found, the migration will add the flag to all `configureTestingModule` and `withModule` calls instead.

IgorMinar reviewed (approved):
LGTM except that docs are missing, can you please work with Joey and Dave to create a one page migration guide similar to those we created in the past: https://angular.io/guide/migration-legacy-message-id

I've seem some tweets over the weekend of people confused about what this teardown feature was all about, so we need to explain of the goal of the migration as well as different knobs we provided for people to smoothly migrate existing test suites over this breaking change. thanks

alxhub reviewed (approved):
Caretaker: this requires a g3patch currently (cl/400842116). Please coordinate and ensure these land together.

dylhunn commented:
This PR was merged into the repository by commit 81c7eb813c27f08d2d640f34e165a1b5e487bac2.
